File: de.unistuttgart.iste.sqa.mpw.hamstersimulator.cpp/core/src/impl/GameViewModelImpl.cpp
#include "GameViewModelImpl.h"

#include "CollectionHelpers.hpp"

#include "ViewModelRow.h"

namespace viewmodel {

GameViewModelImpl::GameViewModelImpl() = default;

std::shared_ptr<viewmodel::ViewModelCell> GameViewModelImpl::getCellAt(
		int rowIndex, int columnIndex) {
    auto row = collectionhelpers::get_at(getRows(), rowIndex);
    auto cell = collectionhelpers::get_at(row->get()->getCells(), columnIndex);
    return cell.value();
}

void GameViewModelImpl::init(mpw::Size size) {
    setSize(size);
    for (int rowIndex = 0; rowIndex < size.getRowCount(); rowIndex++) {
        auto row = std::make_shared<ViewModelRow>();
        addToRows(row);
        for (int columnIndex = 0; columnIndex < size.getColumnCount(); columnIndex++) {
            auto cell = std::make_shared<ViewModelCell>();
            cell->setLocation(mpw::Location::from(columnIndex, rowIndex));
            row->addToCells(cell);
        }
    }
}

}

File: simulators/de.unistuttgart.iste.sqa.mpw.framework.java/core/src/main/java/de/unistuttgart/iste/sqa/mpw/framework/viewmodel/impl/GameViewModelImpl.java
package de.unistuttgart.iste.sqa.mpw.framework.viewmodel.impl;

import de.unistuttgart.iste.sqa.mpw.framework.datatypes.Location;
import de.unistuttgart.iste.sqa.mpw.framework.datatypes.Size;
import de.unistuttgart.iste.sqa.mpw.framework.viewmodel.GameViewModel;
import de.unistuttgart.iste.sqa.mpw.framework.viewmodel.ViewModelCell;
import de.unistuttgart.iste.sqa.mpw.framework.viewmodel.ViewModelRow;

public class GameViewModelImpl extends GameViewModel {

	@Override
	public ViewModelCell getCellAt(final int row, final int column) {
		if (row >= 0 && row < getSize().getRowCount()
				&& column >= 0 && column < getSize().getColumnCount()) {
			return getRows().get(row).getCells().get(column);
		}
		return null;
	}

	@Override
	public void init(final Size size) {
		setSize(size);
		for (int rowIndex = 0; rowIndex < size.getRowCount(); rowIndex++) {
			final var row = new ViewModelRow();
			addToRows(row);
			createCellsForRow(row);
		}
	}

	private void createCellsForRow(final ViewModelRow row) {
		final int rowIndex = getRows().size()-1;
		final Size size = getSize();
		for (int columnIndex = 0; columnIndex < size.getColumnCount(); columnIndex++) {
			final var cell = new ViewModelCell();
			cell.setLocation(Location.from(columnIndex, rowIndex));
			row.addToCells(cell);
		}
	}

}
